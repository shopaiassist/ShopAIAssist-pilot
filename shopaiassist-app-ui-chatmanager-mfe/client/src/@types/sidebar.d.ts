/**
 * Enumeration of available sorting options for lists.
 */
export type SortOptions = 'by_name' | 'by_date' | 'by_type';

/**
 * Defines the type of items, which can be either a folder or a chat.
 */
export type ItemType = 'folder' | 'chat';

/**
 * Defines the structure of a tree item.
 * @interface
 * @property {string} name - The name of the tree item.
 * @property {string} treeItemId - The unique identifier of the tree item.
 * @property {ItemType} type - The type of the item, distinguishing between chat and folder.
 * @property {Date} createdAt - The creation date of the tree item.
 * @property {Date} updatedAt - The update date of the tree item.
 * @property {string} [parentId] - Optional. The ID of the parent folder, if any.
 */
interface TreeItem {
  name: string;
  treeItemId: string;
  type: ItemType;
  createdAt: Date;
  updatedAt: Date;
  fileCollectionId: string;
  parentId?: string;
}

/**
 * Extends TreeItem to define the structure of a folder item, adding folder-specific properties.
 * @interface
 * @property {string} [matterId] - Optional. A external matterId for the folder.
 * @property {string} [description] - Optional. A description of the folder's contents or purpose.
 * @property {string} [isArchived] - Optional. A boolean for if the folders is archived.
 */
interface FolderItem extends TreeItem {
  matterId?: string;
  description?: string;
  isArchived?: boolean;
}
