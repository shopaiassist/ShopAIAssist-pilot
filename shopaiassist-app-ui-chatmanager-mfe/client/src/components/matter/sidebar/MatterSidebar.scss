@import "../../../utils/styles/imports";

$headerHeight: 3.75rem;

.matter-sidebar-container {
   display: flex;
   width: 21rem;
   z-index: map-get($z-index-map, "default");

   &.compact-mode {
      position:absolute;
      width: 100%;
   }

   &.full-mode {
      position:relative;
      width: 21rem;
      max-width: 100%;
   }

   .matter-sidebar-nav-header {
      display: flex;
      padding: 0.5rem;
      justify-content: space-between;
      align-items: center;
      align-self: stretch;
      min-height: 3.688rem;
      box-sizing: border-box;
      width: 100%;
      border: rem-calc(1);
      border-bottom: $saf-line-width-thin solid $saf-color-line-subtle;
      border-left: $saf-line-width-thin solid $saf-color-line-subtle;

      .matter-sidebar-nav-back {
         display: flex;
         gap: 0.5rem;
      }
   }

   .matter-sidebar {
      display: flex;
      flex-direction: column;
      align-items: center;
      box-shadow: $saf-drop-shadow-level-3;
      height: calc(100vh - $headerHeight);
      width: 21rem;
      background-color: $saf-color-background-subtle;

      &.compact-mode {
         width: 100%;
      }

      &.full-mode {
         width: 21rem;
         max-width: 100%;
      }

      .matter-sidebar-folder-name {
         padding: 0.5rem;
         margin: 0.75rem;
         display: flex;
         align-self: flex-start;
      }

      .matter-sidebar-header {
         display: flex;
         flex-direction: row;
         justify-content: space-between;
         width: 95%;
         padding: 0.5rem;
      }

      .matter-sidebar-footer {
         display: flex;
         width: 100%;
      }
   }
}