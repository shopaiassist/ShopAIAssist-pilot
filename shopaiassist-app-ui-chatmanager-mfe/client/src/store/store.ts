import { Action, AsyncThunkPayloadCreator, configureStore, ThunkAction } from '@reduxjs/toolkit';

import userReducer from './userSlice';

export const reducer = {
  user: userReducer
};

export const store = configureStore({
  reducer
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;

export interface AppThunkApiConfig {
  dispatch: AppDispatch;
  rejectValue: string;
  state: RootState;
}
export type AppAsyncThunk<P, R> = AsyncThunkPayloadCreator<R, P, AppThunkApiConfig>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;
