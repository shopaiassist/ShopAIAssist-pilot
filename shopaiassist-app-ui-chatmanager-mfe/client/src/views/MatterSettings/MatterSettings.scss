@import "../../utils/styles/imports";

@mixin flex-col-with-gap {
    display: flex;
    flex-direction: column;
    gap: 1rem;
}

.matter-settings-container {
    padding: $saf-spacing-6;
    
    .matter-settings-fields {
        @include flex-col-with-gap;
        padding: $saf-spacing-6 0;
        max-width: rem-calc(450);
    }

    .matter-settings-bottom {
        @include flex-col-with-gap;
        padding-top: $saf-spacing-6;
    }

    & saf-button {
        width: fit-content;
    }
}