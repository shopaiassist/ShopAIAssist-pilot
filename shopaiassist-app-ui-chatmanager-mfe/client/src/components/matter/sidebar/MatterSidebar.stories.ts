// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
import MatterSidebar from './MatterSidebar';
import { Meta, StoryObj } from '@storybook/react';

const meta = {
  title: 'Matter/Sidebar',
  component: MatterSidebar,
  parameters: {
    layout: 'padded'
  },
  tags: ['autodocs']
} satisfies Meta<typeof MatterSidebar>;

export default meta;

export const Default: StoryObj = { args: {} };
