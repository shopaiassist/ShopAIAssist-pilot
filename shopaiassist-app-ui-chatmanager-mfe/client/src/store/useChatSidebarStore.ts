import { create } from 'zustand';
import { devtools } from 'zustand/middleware';
import axios from 'axios';
import { debounce } from 'lodash';
import { FolderItem, TreeItem } from '../@types/sidebar';
import { Endpoints } from '../utils/api';
import { UserAuth } from 'react';
import { useMfeContextUtilsStore } from './useMfeContextUtilsStore';
import i18nInstance from '../utils/i18n';
import { useBreadCrumbStore } from './useBreadcrumbStore';
import { GENERATE_CHAT_NAME_WAIT_TIME } from '../utils/constatnts';
import { UserProfile } from 'react/dist/types/auth';

export enum SortOptions {
  BY_DATE = 'by_date',
  BY_NAME = 'by_name',
  BY_TYPE = 'by_type'
}

export const ITEM_TYPE = {
  CHAT: 'chat',
  FOLDER: 'folder'
} as const;

/**
 * Defines the structure and types of the state managed in the ChatSidebarStore.
 * @interface ChatSidebarState
 * @property {UserAuth|undefined} userAuth - The user authentication state.
 * @property {(TreeItem|FolderItem)[]|undefined} chatsAndFolders - The list of chats and folders.
 * @property {string} sortType - The current sorting type for displaying the chats and folders.
 * @property {FolderItem|undefined} activeFolder - The currently active folder.
 * @property {TreeItem|undefined} activeChat - The currently active chat.
 * @property {Function} fetchChatList - Function to fetch the list of chats and folders based on the current sort type.
 * @property {Function} deleteChat - Function to delete a chat by its identifier.
 * @property {Function} createChat - Function to create a new chat.
 * @property {Function} updateChatName - Function to update the name of a chat.
 * @property {Function} setSortType - Function to update the sorting type and fetch the updated list.
 * @property {Function} setActiveFolder - Function to set the active folder and fetch its chats.
 * @property {Function} setActiveChat - Function to set the active chat.
 * @property {Function} setUserAuth - Function to set the user authentication state.
 */
export interface ChatSidebarState {
  userAuth: UserAuth | undefined;
  chatsAndFolders?: (TreeItem | FolderItem)[];
  sortType: string;
  activeFolder?: FolderItem;
  activeChat?: TreeItem;
  archivedMattersView?: boolean;
  isNameUpdate: boolean;
  isMatterSettingsActive?: boolean;
  isMatterFilesActive: boolean;
  isLoading: boolean;
  activeUser: UserProfile;
  fetchChatList: () => void;
  deleteChat: (chatId: string) => void;
  createChat: (message: string, chatFolderIdentifier?: string) => Promise<string>;
  updateChatName: (chatId: string, newChatName: string) => void;
  generateChatName: (chatId: string, chatHistory: { role: string, content: string | undefined }[]) => void;
  setSortType: (newSortType: string) => void;
  setActiveFolder: (newActiveFolder?: FolderItem) => void;
  setActiveChat: (newActiveChat?: TreeItem) => void;
  setUserAuth: (userAuth: UserAuth) => void;
  toggleArchivedMattersView: () => void;
  setIsMattersettingsActive: (isMatterSettingsActive: boolean) => void;
  refreshActiveFolder: () => Promise<void>;
  setIsMatterFilesActive: (isMatterFilesActive: boolean) => void;
  setIsNameUpdate: (isNameUpdateNewValue: boolean) => void;
  setActiveUser: (activeUser: UserProfile) => void;
}

/**
 * The base URL constructor for API endpoints, utilizes the environment variable for the app's domain.
 * @param {string} endpoint - The API endpoint to be appended to the base domain.
 * @returns {string} The full URL for the API call.
 */
const BASE_URL = (endpoint: string) => `${process.env.APP_DOMAIN}/api/${endpoint}`;

// const SAMPLE_CHATS: TreeItem[] = [
//   {
//     treeItemId: 'c91fed15-2e50-45eb-a59a-666666667777',
//     name: '--A - CREATE NEW CHAT--',
//     createdAt: new Date('2024-03-04T18:43:32.194686+00:00'),
//     updatedAt: new Date(),
//     type: 'chat',
//     fileCollectionId: ''
//   },
//   {
//     treeItemId: 'c91fed15-2e50-45eb-a59a-cebe44e2bf7b',
//     name: 'ZZZConversation Naming Assistant Capabilities',
//     createdAt: new Date('2024-03-04T18:43:32.194686+00:00'),
//     updatedAt: new Date('2024-03-04T18:43:32.194686+00:00'),
//     type: 'chat',
//     fileCollectionId: ''
//   },
//   {
//     treeItemId: '5a44626f-b1b8-4943-acc3-62651e2c39a3',
//     name: 'PDF File Upload: 070823vanliere (1)',
//     createdAt: new Date('2024-03-05T16:32:52.361758+00:00'),
//     updatedAt: new Date('2024-03-05T16:32:52.361758+00:00'),
//     type: 'chat',
//     fileCollectionId: ''
//   }
// ];

/**
 * Zustand store to manage the sidebar state, including chats and folders and related actions.
 * Uses axios for HTTP requests and lodash's debounce for limiting the rate of API calls.
 */
export const useChatSidebarStore = create<ChatSidebarState>()(
  devtools(
    (set, get) => ({
      userAuth: undefined,
      chatsAndFolders: [],
      sortType: SortOptions.BY_DATE,
      activeFolder: undefined,
      activeChat: undefined,
      isMatterSettingsActive: undefined,
      isMatterFilesActive: false,
      isNameUpdate: false,
      isLoading: false,
      activeUser: {
        email: '',
        firstName: '',
        lastName: '',
        username: '',
        registrationKey: '',
        region: '',
        id: ''
      },
      /**
       * Fetches the list of chats and folders based on the current sort type.
       * Uses a debounced function to limit the rate of API calls.
       * Sets isLoading to true during the fetch operation.
       */
      fetchChatList: debounce(async () => {
        // Set loading state to true when starting fetch
        set({ isLoading: true });
        try {
          const res = await axios.get(BASE_URL(Endpoints.LIST), {
            params: {
              sortType: get().sortType,
              parentId: get().activeFolder?.treeItemId,
              onlyArchivedMatters: get().archivedMattersView,
              uid: get().activeUser.id
            }
          });
          // Validate API response and update state with new data
          const chatsAndFolders = Array.isArray(res.data) ? res.data : [];
          if (!Array.isArray(res.data)) {
            console.error('API returned unexpected data structure for fetchChatList:', res.data);
          }
          set({ chatsAndFolders, isLoading: false });
        } catch (error) {
          // Make sure to set loading to false even if there's an error
          console.error('Error fetching chat list:', error);
          set({ isLoading: false });
          useMfeContextUtilsStore.getState().triggerAlert(i18nInstance.t('ERROR_MESSAGE'), 'error');
        }
      }),
      /**
       * Deletes a chat by its identifier and refreshes the chat list.
       * Uses a debounced function to limit the rate of API calls.
       *
       * @param {string} chatId - The identifier of the chat to be deleted.
       */
      deleteChat: debounce(async (chatId: string) => {
        const deleteChatResult = await axios.delete(BASE_URL(Endpoints.CHAT_BY_ID), {
          data: { id: chatId, user: get().activeUser }
        });
        if (deleteChatResult.status === 200) {
          useMfeContextUtilsStore.getState().triggerAlert(i18nInstance.t('CHAT_DELETE_SUCCESS'), 'success');
        } else {
          useMfeContextUtilsStore.getState().triggerAlert(i18nInstance.t('ERROR_MESSAGE'), 'error');
        }
        get().setActiveChat();
        get().fetchChatList();
      }),

      /**
       * Creates a new chat and returns the chat ID.
       *
       * @param {string} [chatFolderIdentifier] - The identifier of the chat folder.
       * @returns {Promise<string>} The new chat ID.
       */
      createChat: async (message: string | undefined, chatFolderIdentifier?: string): Promise<string> => {
        const newChatData = await axios.post(
          BASE_URL(Endpoints.LIST),
          {
            parentId: chatFolderIdentifier,
            user: get().activeUser
          },
          { withCredentials: true }
        );

        get().setActiveChat(newChatData.data.newChat as TreeItem);
        console.info(`Created new chat with ID: ${newChatData.data.newChat.treeItemId}`);
        get().generateChatName(newChatData.data.newChat.treeItemId, [{ role: 'user', content: message ?? ''}] );
        return newChatData.data.newChat.treeItemId as string;
      },
      /**
       * Generates a new name for a chat and updates the chat name.
       * Uses a debounced function to limit the rate of API calls.
       *
       * @param {string} chatId - The identifier of the chat to be updated.
       */
      generateChatName: debounce(async (chatId: string, chatHistory: { role: string, content: string | undefined }[]) => {
        setTimeout(async () => {
          try {
            const generatedName = await axios.post(BASE_URL('generate-name'), { chatId, chat_history: chatHistory });

            if (generatedName.data.newName) {
              const activeChat = get().activeChat;

              if (activeChat !== undefined && activeChat.treeItemId === chatId) {
                set({ isNameUpdate: true });
                get().setActiveChat({ ...activeChat, name: generatedName.data.newName });
                
                // wait for a moment then update the chat list.
                setTimeout(() => {
                  get().fetchChatList();
                }, GENERATE_CHAT_NAME_WAIT_TIME);
              }
            }
          } catch (error) {
            console.error('Error generating chat name:', error);
            useMfeContextUtilsStore.getState().triggerAlert(i18nInstance.t('ERROR_MESSAGE'), 'error');
          }          
        }, GENERATE_CHAT_NAME_WAIT_TIME);
      }),
      /**
       * Updates the name of a chat and refreshes the chat list.
       * Uses a debounced function to limit the rate of API calls.
       *
       * @param {string} chatId - The identifier of the chat to be updated.
       * @param {string} newChatName - The new name for the chat.
       */
      updateChatName: debounce(async (chatId: string, newChatName: string) => {
        await axios
          .post(BASE_URL(`chat/${chatId}/rename`), { id: chatId, updates: { name: newChatName }, user: get().activeUser })
          .then(() => {
            useBreadCrumbStore.getState().updateBreadcrumb(chatId, newChatName);
            set({ isNameUpdate: true });
            const activeChat = get().activeChat;
            if (activeChat !== undefined && activeChat.treeItemId === chatId) {
              get().setActiveChat({ ...activeChat, name: newChatName });
            }
            get().fetchChatList();
          });
      }),
      /**
       * Updates the sorting type and fetches the updated list of chats and folders.
       *
       * @param {string} newSortType - The new sorting type.
       */
      setSortType: (newSortType: string) => {
        set({ sortType: newSortType });
        get().fetchChatList();
      },
      /**
       * Sets the active folder and fetches the list of chats within the active folder.
       *
       * @param {FolderItem|undefined} newActiveFolder - The new active folder.
       */
      setActiveFolder: (newActiveFolder?: FolderItem) => {
        if (newActiveFolder) {
          useBreadCrumbStore.getState().addBreadcrumb({
            icon: 'folder',
            label: newActiveFolder.name,
            href: 'javascript:;',
            onClick: () => {
              get().setActiveChat();
            },
            testId: newActiveFolder.treeItemId
          });
          get().fetchChatList();
        } else {
          const prevFolder = get().activeFolder;
          if (prevFolder) {
            useBreadCrumbStore.getState().removeBreadcrumb(prevFolder.treeItemId);
          }

          get().setIsMatterFilesActive(false);
          get().setIsMattersettingsActive(false);
          get().fetchChatList();
        }
        set({ activeFolder: newActiveFolder });
      },
      /**
       * Sets the active chat.
       *
       * @param {TreeItem|undefined} newActiveChat - The new active chat.
       */
      setActiveChat: (newActiveChat?: TreeItem) => {
        if (newActiveChat) {
          useBreadCrumbStore.getState().addBreadcrumb({
            icon: 'message-lines',
            label: newActiveChat.name,
            href: 'javascript:;',
            onClick: () => {},
            testId: newActiveChat.treeItemId
          });
        }
        const prevActiveChat = get().activeChat;
        if (prevActiveChat) {
          useBreadCrumbStore.getState().removeBreadcrumb(prevActiveChat.treeItemId);
        }

        get().setIsMattersettingsActive(false);
        get().setIsMatterFilesActive(false);
        set({ activeChat: newActiveChat });
      },
      /**
       * Sets the user authentication state.
       *
       * @param {UserAuth} userAuth - The user authentication state.
       */
      setUserAuth: (userAuth: UserAuth) => {
        set({ userAuth });
      },
      /**
       * Toggle archived matters view
       */
      toggleArchivedMattersView: () => {
        set({ archivedMattersView: !get().archivedMattersView });

        if (get().archivedMattersView) {
          useBreadCrumbStore.getState().addBreadcrumb({
            icon: 'box-archive',
            label: 'Archived Matters',
            href: 'javascript:;',
            onClick: () => {
              get().setActiveFolder(undefined);
            },
            testId: 'archived_matters'
          });
        } else {
          useBreadCrumbStore.getState().removeBreadcrumb('archived_matters');
        }

        get().fetchChatList();
        /** Sets the matter settings active state.
         *
         * @param {boolean} isMatterSettingsActive - The value to be set.
         */
      },
      setIsMattersettingsActive: (isMatterSettingsActive: boolean) => {
        if (isMatterSettingsActive) {
          useBreadCrumbStore.getState().addBreadcrumb({
            icon: 'gear',
            label: 'Matter settings',
            href: 'javascript:;',
            onClick: () => {},
            testId: 'matter_settings'
          });
        } else {
          useBreadCrumbStore.getState().removeBreadcrumb('matter_settings');
        }
        set({ isMatterSettingsActive });
      },
      /**
       * Sets the matter settings active state.
       *
       * @param {boolean} isMatterSettingsActive - The value to be set.
       */
      setIsMatterFilesActive: (isMatterFilesActive: boolean) => {
        if (isMatterFilesActive) {
          useBreadCrumbStore.getState().addBreadcrumb({
            icon: 'files',
            label: 'Matter files',
            href: 'javascript:;',
            onClick: () => {},
            testId: 'matter_files'
          });
        } else {
          useBreadCrumbStore.getState().removeBreadcrumb('matter_files');
        }
        set({ isMatterFilesActive });
      },
      refreshActiveFolder: async () => {
        const activeFolder = get().activeFolder;
        if (!activeFolder) return;

        try {
          const res = await axios.get(BASE_URL(Endpoints.LIST), {
            params: {
              sortType: get().sortType,
              parentId: get().activeFolder?.treeItemId,
              onlyArchivedMatters: get().archivedMattersView,
              uid: get().activeUser.id
            }
          });
          
          const chatsAndFolders = res.data;
          if (Array.isArray(chatsAndFolders)) {
            const folderToRefresh = chatsAndFolders.find((item) => item.treeItemId === activeFolder.treeItemId);
            if (folderToRefresh) {
              set({ activeFolder: folderToRefresh });
            }
          } else {
            console.error('API returned unexpected data structure for refreshActiveFolder:', chatsAndFolders);
            useMfeContextUtilsStore.getState().triggerAlert(i18nInstance.t('ERROR_MESSAGE'), 'error');
          }
        } catch (error) {
          console.error('Error refreshing active folder:', error);
          useMfeContextUtilsStore.getState().triggerAlert(i18nInstance.t('ERROR_MESSAGE'), 'error');
        }
      },
      setIsNameUpdate(isNameUpdateNewValue) {
        set({ isNameUpdate: isNameUpdateNewValue });
      },
      setActiveUser(activeUser: UserProfile) {
        set({ activeUser });
      }
    }),
    { name: 'useChatSidebarStore' }
  )
);
