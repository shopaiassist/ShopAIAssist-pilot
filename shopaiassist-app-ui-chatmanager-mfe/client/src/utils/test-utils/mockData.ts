import { TreeItem, FolderItem } from '../../@types/sidebar';

export const mockTree: (TreeItem | FolderItem)[] = [
  {
    treeItemId: '9a4144e0-65ea-4f6c-8cd5-43a0a6d1e634',
    name: 'Test Folder For Created At 2',
    type: 'folder',
    createdAt: new Date('2024-05-06T15:26:57.992Z'),
    updatedAt: new Date('2024-05-06T15:26:57.992Z'),
    fileCollectionId: 'whatever'
  },
  {
    treeItemId: '5aa13d2b-3f04-4a91-a84b-bf214420b47c',
    name: 'Test Folder For Created At 2',
    type: 'folder',
    createdAt: new Date('2024-05-06T15:19:07.430Z'),
    updatedAt: new Date('2024-05-06T15:19:07.430Z'),
    fileCollectionId: 'whatever'
  },
  {
    type: 'folder',
    name: 'New Folder 100',
    description: 'This folder is used for whatever, its created via postman.',
    matterId: 'Whatever Folder 100',
    treeItemId: '4c484fb3-faf6-4fe8-9930-410772f141d7',
    createdAt: new Date('2024-04-25T19:21:07.787Z'),
    updatedAt: new Date('2024-04-25T19:21:07.787Z'),
    fileCollectionId: 'whatever'
  },
  {
    createdAt: new Date('2024-03-01T17:00:33.564Z'),
    treeItemId: 'e35607c7-be8f-4172-a3cd-ae7e39bc2b96',
    name: 'A chat named desire',
    type: 'chat',
    updatedAt: new Date('2024-05-06T18:09:49.245Z'),
    fileCollectionId: 'whatever'
  },
  {
    createdAt: new Date('2024-02-23T16:30:33.080Z'),
    updatedAt: new Date('2024-02-23T16:30:33.080Z'),
    treeItemId: '8c1b78b3-a798-4557-871a-bff6c8ac804e',
    name: 'new2',
    type: 'folder',
    fileCollectionId: 'whatever'
  },
  {
    createdAt: new Date('2024-02-23T16:30:17.078Z'),
    updatedAt: new Date('2024-02-23T16:30:17.078Z'),
    treeItemId: 'b39f37a3-6781-4445-8573-77658d9ee7ae',
    name: 'new1',
    type: 'folder',
    fileCollectionId: 'whatever'
  },
  {
    createdAt: new Date('2024-02-23T16:26:17.254Z'),
    updatedAt: new Date('2024-02-23T16:26:17.254Z'),
    treeItemId: '9bc3c583-a6e4-41cc-92f1-27e1be3ba4f6',
    name: 'newnewenewnew',
    type: 'folder',
    fileCollectionId: 'whatever'
  },
  {
    createdAt: new Date('2024-02-23T16:25:50.258Z'),
    updatedAt: new Date('2024-02-23T16:25:50.258Z'),
    treeItemId: '2ec6bba4-6dd0-4190-a015-b58e6ee4021c',
    name: '12346123445',
    type: 'folder',
    fileCollectionId: 'whatever'
  },
  {
    createdAt: new Date('2024-02-21T20:51:41.137Z'),
    updatedAt: new Date('2024-02-21T20:51:41.137Z'),
    treeItemId: 'e4a9bc6a-cfb1-48ac-bc6e-1b6f135147c0',
    name: 'Chat Name 2',
    type: 'chat',
    fileCollectionId: 'whatever'
  },
  {
    createdAt: new Date('2024-02-20T19:47:03.738Z'),
    updatedAt: new Date('2024-02-20T19:47:03.738Z'),
    treeItemId: 'ce2422ea-b573-4c71-b569-fa2c79a5e50b',
    name: 'B matter with B name',
    type: 'folder',
    fileCollectionId: 'whatever'
  },
  {
    createdAt: new Date('2024-02-20T19:44:31.642Z'),
    updatedAt: new Date('2024-02-20T19:44:31.642Z'),
    treeItemId: '148f72ac-b903-451c-88f2-ed1282c6acb1',
    name: 'Zmatter with Z name',
    matterId: 'A matter with A name',
    type: 'folder',
    fileCollectionId: 'whatever'
  },
  {
    createdAt: new Date('2024-02-15T18:35:41.269Z'),
    updatedAt: new Date('2024-02-15T18:35:41.269Z'),
    treeItemId: 'e82e81e6-68c8-4bd1-b67c-316dc4105d18',
    name: '123',
    matterId: 'Matter with a really REALLY long name, im mean like REALLY long',
    type: 'folder',
    fileCollectionId: 'whatever'
  }
];
