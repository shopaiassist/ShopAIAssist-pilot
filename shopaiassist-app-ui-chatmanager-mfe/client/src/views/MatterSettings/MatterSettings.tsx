import { Icon, Button, Text, Textarea, Textfield } from '@on/core-components/react';
import './MatterSettings.scss';
import { useMatterSettings } from '../../hooks/useMatterSettings';
import { UpdateType } from '../../hooks/useMatterDetailsDialog';
import Translate from '../../components/i18n/Translate';
import { useTranslation } from 'react-i18next';

/**
 * The Matter Settings page c|omponent
 * This component handles edit, archive and delete actions for the current folder
 *
 * @returns {JSX.Element} The rendered MatterSettings component.
 */
const MatterSettings = () => {
  const { t } = useTranslation();
  const [
    activeFolder,
    matterName,
    matterId,
    matterDescription,
    updateMatterDetails,
    onDeleteMatter,
    onUpdateMatter,
    onArchiveMatter,
    onUnarchiveMatter
  ] = useMatterSettings();

  return (
    <div className="matter-settings-container">
      <Text appearance="heading-2xl" className="matter-settings-title">
        <Translate tKey="MATTER_SETTINGS" />
      </Text>

      <div className="matter-settings-fields">
        <Text appearance="body-default-md">{`* ${t('INDICATES_REQUIRED_FIELD')}`}</Text>
        <Textfield
          data-testid="matter-settings-name-field"
          label={t('NAME')}
          maxlength={128}
          value={matterName}
          required={true}
          invalid={!matterName}
          validationMessage={t('REQUIRED_FIELD')}
          onInput={(e) => {
            updateMatterDetails(UpdateType.NAME, e.target.value);
          }}
        />
        <Textarea
          data-testid="matter-settings-description-field"
          label={t('DESCRIPTION')}
          maxlength={2000}
          resize="vertical"
          value={matterDescription}
          onInput={(e) => {
            updateMatterDetails(UpdateType.DESCRIPTION, e.target.value);
          }}
        />
        <Textfield
          data-testid="matter-settings-id-field"
          label={t('MATTER_ID')}
          maxlength={128}
          value={matterId}
          onInput={(e) => {
            updateMatterDetails(UpdateType.ID, e.target.value);
          }}
        />

        <Button disabled={!matterName.length} onClick={onUpdateMatter} data-testid="matter-settings-update">
          <Translate tKey="UPDATE_FOLDER" />
        </Button>
      </div>

      <div className="matter-settings-bottom">
        {!activeFolder.isArchived && (
          <Button appearance="secondary" onClick={onArchiveMatter} data-testid="matter-settings-archive">
            <Icon slot="start" iconName="box-archive" />
            <Translate tKey="ARCHIVE_FOLDER" />
          </Button>
        )}
        {activeFolder.isArchived && (
          <Button appearance="secondary" onClick={onUnarchiveMatter} data-testid="matter-settings-unarchive">
            <Icon slot="start" iconName="box-archive" />
            <Translate tKey="UNARCHIVE_FOLDER" />
          </Button>
        )}
        <Button appearance="secondary" onClick={onDeleteMatter} data-testid="matter-settings-delete">
          <Icon slot="start" iconName="trash-can" />
          <Translate tKey="DELETE_MATTER" />
        </Button>
      </div>
    </div>
  );
};

export default MatterSettings;
