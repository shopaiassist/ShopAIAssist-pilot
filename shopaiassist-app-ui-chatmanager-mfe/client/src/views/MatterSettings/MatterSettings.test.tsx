import axios from 'axios';
import { act, fireEvent, renderWithI18n, snapshot, waitFor } from '../../../jest-utils';
import { useChatSidebarStore } from '../../store/useChatSidebarStore';
import MatterSettings from './MatterSettings';
import { mockTree } from '../../utils/test-utils/mockData';
import { FolderItem } from '../../@types/sidebar';

const getMockedComponent = () => {
  return <MatterSettings />;
};

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;
const mockedMatter: FolderItem = {
  treeItemId: 'fake-item-id',
  name: 'fakeMatter',
  createdAt: new Date(),
  updatedAt: new Date(),
  type: 'folder',
  description: 'This is a fake matter',
  matterId: 'fake-custom-matter-id',
  fileCollectionId: '1234'
};

describe.only('MatterSettings Tests', () => {
  const store = useChatSidebarStore.getState();

  beforeEach(() => {
    useChatSidebarStore.setState(store, true);
    store.setActiveFolder(mockedMatter);
    jest.clearAllMocks();
  });

  describe('render', () => {
    it('should render', () => {
      snapshot(getMockedComponent());
    });
  });

  describe('user action', () => {
    it('should call to update a matter', async () => {
      store.setActiveFolder(mockedMatter);
      store.setIsMattersettingsActive(true);
      mockedAxios.get.mockResolvedValue(mockTree);
      mockedAxios.patch.mockResolvedValue({ status: 200 });
      const component = getMockedComponent();
      const wrapper = renderWithI18n(component);

      const nameField = wrapper.getByTestId('matter-settings-name-field');
      expect(nameField).not.toBeDisabled();
      fireEvent.input(nameField, { target: { value: 'Changed Name' } });

      const descField = wrapper.getByTestId('matter-settings-description-field');
      expect(descField).not.toBeDisabled();
      fireEvent.input(descField, { target: { value: 'Changed Description' } });

      const matterIdField = wrapper.getByTestId('matter-settings-id-field');
      expect(matterIdField).not.toBeDisabled();
      fireEvent.input(matterIdField, { target: { value: 'Changed Matter ID' } });

      const updateButton = wrapper.getByTestId('matter-settings-update');
      expect(updateButton).not.toBeDisabled();
      updateButton.click();

      await act(() => waitFor(() => wrapper.rerender(component)));
      expect(mockedAxios.patch).toHaveBeenCalledWith('jest-test-domain/api/folders', {
        id: mockedMatter.treeItemId,
        updates: { description: 'Changed Description', matterId: 'Changed Matter ID', name: 'Changed Name' }
      });
    });
  });
});
