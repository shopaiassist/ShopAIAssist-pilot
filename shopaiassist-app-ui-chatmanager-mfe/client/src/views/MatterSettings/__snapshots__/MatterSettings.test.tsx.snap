// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MatterSettings Tests render should render 1`] = `
<div
  class="matter-settings-container"
>
  <span
    appearance="heading-2xl"
    class="matter-settings-title"
  >
    Matter settings
  </span>
  <div
    class="matter-settings-fields"
  >
    <span
      appearance="body-default-md"
    >
      * indicates required field
    </span>
    <textarea
      data-testid="matter-settings-name-field"
      label="Name"
      maxlength="128"
      required=""
      validationmessage="This field is required"
    >
      fakeMatter
    </textarea>
    <textarea
      data-testid="matter-settings-description-field"
      label="Description"
      maxlength="2000"
      resize="vertical"
    >
      This is a fake matter
    </textarea>
    <textarea
      data-testid="matter-settings-id-field"
      label="Matter ID"
      maxlength="128"
    >
      fake-custom-matter-id
    </textarea>
    <button
      data-testid="matter-settings-update"
    >
      Update matter
    </button>
  </div>
  <div
    class="matter-settings-bottom"
  >
    <button
      appearance="secondary"
      data-testid="matter-settings-archive"
    >
      <div
        iconname="box-archive"
        slot="start"
      />
      Archive matter
    </button>
    <button
      appearance="secondary"
      data-testid="matter-settings-delete"
    >
      <div
        iconname="trash-can"
        slot="start"
      />
      Delete matter
    </button>
  </div>
</div>
`;
